src/items.rs

```diff
@@ -174,22 +235,82 @@ pub enum Instance {
     },
 }
 
+impl Display for Instance {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Instance::Cnf { num_vars, clauses } => {
+                Instance::fmt_cnf(f, *num_vars, clauses)
+            }
+            Instance::Sat {
+                num_vars,
+                extensions,
+                formula,
+            } => Instance::fmt_sat(f, *num_vars, extensions, formula),
+        }
+    }
+}
+
 impl Instance {
     /// Creates a new SAT instance for `.cnf` files with given clauses.
     pub fn cnf(num_vars: u64, clauses: Vec<Clause>) -> Instance {
         Instance::Cnf {
-            num_vars: num_vars,
+            num_vars,
             clauses: clauses.into_boxed_slice(),
         }
     }
 
-    /// Creates a new SAT instance for `.sat` files with given extensions and an underlying formula.
-    pub fn sat(num_vars: u64, extensions: Extensions, formula: Formula) -> Instance {
+    /// Creates a new SAT instance for `.sat` files with given extensions and
+    /// an underlying formula.
+    pub fn sat(
+        num_vars: u64,
+        extensions: Extensions,
+        formula: Formula,
+    ) -> Instance {
         Instance::Sat {
-            num_vars: num_vars,
-            extensions: extensions,
-            formula: formula,
+            num_vars,
+            extensions,
+            formula,
+        }
+    }
+
+    fn fmt_sat(
+        f: &mut std::fmt::Formatter<'_>,
+        num_vars: u64,
+        extensions: &Extensions,
+        formula: &Formula,
+    ) -> std::fmt::Result {
+        writeln!(f, "p {} {}", extensions, num_vars)?;
+        writeln!(f, "{}", formula)
+    }
+
+    fn fmt_cnf(
+        f: &mut std::fmt::Formatter<'_>,
+        num_vars: u64,
+        clauses: &[Clause],
+    ) -> std::fmt::Result {
+        writeln!(f, "p cnf {} {}", num_vars, clauses.len())?;
+        for clause in clauses.iter() {
+            for literal in clause.lits() {
+                write!(f, "{} ", literal)?;
+            }
+            writeln!(f, "0")?;
         }
+        Ok(())
+    }
+
+    /// Creates a SAT or CNF instance, converting it into a String
+    /// `comments` is a list of comments which are inserted into the
+    /// beginning of the resulting String.
+    pub fn serialize<O: core::fmt::Write>(
+        &self,
+        comments: &[String],
```

Comment:
I just noticed this. This would enforce a buffer of memory allocated string objects. More generic was to just use `comments: &[&str]` instead, even if it looks slightly odd. Even better was to use an input iterator but that is a bit more advanced and not needed.

Reply:
Alright, made that change.